### PR TITLE
Dynamo Right Side Extensions Panel

### PR DESCRIPTION
**Pull request by QilongTang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

This PR is aiming to give proposal to new UX using multiple view extensions in Dynamo by prototyping Dynamo right side extensions panel within Dynamo main windows and provide API for extension developer to inject customized UI into Dynamo.

Current progress:
image

TODO (yes, a lot of TODOs)
- [x] Fix UI so that the drag and drop experience is seemless like library view extension on the left
- [x] Propose APIs so library view extension will use that to inject it self
- [x] Provide another view extension example using right side WPF grid
NodeSuggestionInDynamoRightSideExtensionPanel_v2

- [ ] Use Dynamo with multiple view extension open at the same time and test

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner 

### FYIs

@DynamoDS/dynamo

### CODE REVIEW

**Comment by mjkkirschner** [comment]
have you considered adding a tabbed view so that the API adds a new tab with the extension name? That way extensions don't stomp on each other?

**Comment by QilongTang** [comment]
@mjkkirschner It is definitely in my plan, similar to what we are doing in AVP but beyond

**Comment by mjkkirschner** [comment]
@QilongTang whats the plan here, do we have a follow up task to realize this?

**Comment by QilongTang** [comment]
@mjkkirschner I believe what we discussed is:

- @ColinDayOrg  on Dynamo Team side will wrap up the work in https://github.com/DynamoDS/Dynamo/pull/9667

- Extra PR opened as https://github.com/DynamoDS/Dynamo/pull/9706 by @Alvaro so our team have a reference of what event we may change from internal to public which will be in a different PR

- Once https://github.com/DynamoDS/Dynamo/pull/9691 is finalized, work on view inject API for @Alvaro Ortega Pickmans's team to try on

- Dynamo Team to plan for surfacing graph info other than per node running, e.g. callsite info

In short, we need a task to modify stuff based on team's comments in my UX proposal and review/merge this PR. Everything should be useful in this PR from my assessment so far, but we are missing the view injection UI for extensions to use.

File: src/DynamoCoreWpf/Properties/Resources.resx [hunk 1]
```diff
@@ -1949,10 +1949,10 @@ Do you want to install the latest Dynamo update?</value>
     <value>Package Path Added</value>
   </data>
   <data name="PackagePathAutoAddNotificationShortDescription" xml:space="preserve">
-    <value>A library (*.dll, *.ds) was recently imported into Dynamo. Its path was automatically added to "Settings > Manage Node and Package Paths..."</value>
+    <value>A library (*.dll, *.ds) was recently imported into Dynamo. Its path was automatically added to "Settings &gt; Manage Node and Package Paths..."</value>
   </data>
   <data name="PackagePathAutoAddNotificationDetailedDescription" xml:space="preserve">
-    <value>The import path "{0}" was added to "Manage Node and Package Paths". If you want to update or remove this path, please open "Settings > Manage Node and Package Paths..."</value>
+    <value>The import path "{0}" was added to "Manage Node and Package Paths". If you want to update or remove this path, please open "Settings &gt; Manage Node and Package Paths..."</value>
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Properties/Resources.resx]: These changes are generated by VS, should be equivalent.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml [hunk 2]
```diff
@@ -196,24 +196,34 @@
         <Grid.ColumnDefinitions>
             <ColumnDefinition MaxWidth="500"
                               MinWidth="1"
-                              Name="LibraryViewColumn">
-                <ColumnDefinition.Width>
+                              Name="LeftExtensionsViewColumn">
+                <!--<ColumnDefinition.Width>
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml]: This is commented out because I found it unnecessary and lead to broken behavior, but need to run unit tests on this

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml [hunk 3]
```diff
@@ -196,24 +196,34 @@
...
+            </ColumnDefinition>
         </Grid.ColumnDefinitions>
 
         <!--Titlebar-->
         <Border BorderBrush="#FF3F4040"
                 BorderThickness="0,0,0,1"
                 Grid.Row="0"
                 Grid.Column="0"
-                Grid.ColumnSpan="3"
+                Grid.ColumnSpan="5"
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml]: If you guys find it difficult to understand why I made this change, that indicates team need knowledge share with me on WPF UI grid. I have shared why I had to expand the number here with @scottmitchell but would be happy to cover this with team

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml [hunk 4]
```diff
@@ -1478,21 +1490,21 @@
...
                     Width="Auto"
                     Height="Auto"
                     Background="#222222"
                     HorizontalAlignment="Left"
                     VerticalAlignment="Top"
-                    Margin="0,120,0,0">
+                    Margin="0,310,0,0">
             <Grid Background="#353535"
-                  Width="80"
+                  Width="100"
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml]: I made both left and right side button to be 100px wide so they are symmetric

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml [hunk 5]
```diff
@@ -1591,13 +1671,27 @@
...
+        <Grid Height="Auto"
+              Width="Auto"
+              HorizontalAlignment="Stretch"
+              Name="sidebarExtensionsGrid"
+              VerticalAlignment="Stretch"
+              Visibility="Visible"
+              Grid.Row="2"
+              Grid.Column="4"
+              Grid.RowSpan="2">
+            <!--<DataGrid ItemsSource="{Binding Workspaces[0].Nodes}" />-->
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml]: This is a trial to inject chart's data, will remove

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml [hunk 6]
```diff
@@ -196,24 +196,34 @@
         <Grid.ColumnDefinitions>
             <ColumnDefinition MaxWidth="500"
                               MinWidth="1"
-                              Name="LibraryViewColumn">
-                <ColumnDefinition.Width>
+                              Name="LeftExtensionsViewColumn">
+                <!--<ColumnDefinition.Width>
                     <Binding Path="LibraryWidth"
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml]: where is this property set? Is something using it and setting it externally? Now it will do nothing right?

Comment (QilongTang) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml]: Preference Setting if I am correct

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml [hunk 7]
```diff
@@ -1478,21 +1490,21 @@
 
         </Border>
 
-        <!--Collapsed Sidebar-->
+        <!--Collapsed Left Sidebar-->
         <StackPanel Grid.Row="2"
                     Grid.Column="2"
                     Orientation="Vertical"
-                    Name="collapsedSidebar"
+                    Name="collapsedLibrarySidebar"
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml]: I wonder if we should consider these names public APIs?

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml [hunk 8]
```diff
@@ -1544,7 +1556,75 @@
...
         </StackPanel>
+        
+        <!--Collapsed Extension Sidebar-->
+        <StackPanel Grid.Row="2"
+                    Grid.Column="2"
+                    Orientation="Vertical"
+                    Name="collapsedExtensionSidebar"
+                    Width="Auto"
+                    Height="Auto"
+                    Background="#222222"
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml]: is there a dictionary with this color somewhere?

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml (hunk 7 above)
Comment (QilongTang) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml]: You mean so people could access these without looking at our code?

**Comment by QilongTang** [comment]
Addressed all the comments, let me know

**Comment by QilongTang** [comment]
After talking with Team at standup, merging. The actual API for extension view injection are being worked on and will be in a different PR
